doc/bulk_counter/bulk_counter.md

```diff
@@ -0,0 +1,144 @@
+
+##### Bulk Statistic Context
+
+A new structure shall be added to FlexCounter class. 
+
+This structure is created because:
+
+- Meet the signature of sai_bulk_object_get_stats and sai_bulk_object_clear_stats
+- Avoid constructing these information each time collecting statistic. The bulk context shall only be updated under below cases:
+  - New object join counter group. E.g. adding a new port object.
+  - Existing object leave counter group. E.g removing an existing port object.
+
+```cpp
+struct BulkStatsContext
+{
+    bool support = true;
+    std::vector<sai_object_id_t> object_vids;
+    std::vector<sai_object_key_t> object_keys; 
+    std::vector<sai_stat_id_t> counter_ids;
+    std::vector<sai_status_t> get_statuses;
+    std::vector<sai_status_t> clear_statuses;
+    std::vector<uint64_t> counters;
+    std::shared_ptr<sai_stat_capability_list_t> stats_capas;
+};
+```
+- support: indicate if bulk is supported for this object type.
+- object_keys: objects that participate the bulk call. E.g. for port, SAI object id value shall be put into sai_object_key_t structure.
+- counter_ids: SAI statistic IDs that will be queried/cleared by the bulk call.
+- get_statuses: SAI bulk get return value for each object.
+- clear_statuses: SAI bulk clear return value for each object. Only effective if stats mode is read and clear.
+- counters: counter values that will be fill by vendor SAI.
+- stats_capas: stats capability for each statitstic IDs for current object type.
+
+The flow of how to updating bulk context will be discussed in following section.
+
+An instance of BulkStatsContext shall be added to FlexCounter class for each object type.
+
+```cpp
+
+BulkStatsContext m_portBulkContext;
+BulkStatsContext m_priorityGroupBulkContext;
+BulkStatsContext m_queueBulkContext;
+...
+
+```
+
+##### Update Bulk Context
+
+1. New object join counter group.
+
+![Add Object Flow](/doc/bulk_counter/object_join_counter_group.svg).
+
+2. Existing object leave counter group, related data shall be removed from bulk context.
+
+##### Statistic Collect
+
+![Collect Counter Flow](/doc/bulk_counter/counter_collect.svg).
+
+### SAI API
+
+SAI APIs shall be used in this feature:
+
+- sai_bulk_object_get_stats
+- sai_bulk_object_clear_stats
+
+### Configuration and management
+
+N/A
+
+### Warmboot and Fastboot Design Impact
+
+No extra logic on SONiC side is needed to handle warmboot/fastboot.
+
+- As fastboot dealys all counters querying, this feature does not affect fastboot. 
+- For warmboot, it is vendor SAI implementation's responsible to make sure that there must be no error if warmboot starts while bulk API is called.
+
+### Restrictions/Limitations
+
+- Bulk collect attribute value is not supported
+- Buffer pool stats is not support for bulk because different buffer pool may have different stats mode. E.g. pool1 has mode SAI_STATS_MODE_READ, pool2 has mode SAI_STATS_MODE_READ_AND_CLEAR. The new SAI bulk API only allows specify one stats mode.
+
```

Comment:
Any limitation on maximum object number at one bulk get?

Reply:
It depends on vendor implementation. I will add an item to this chapter.